Doris 0.13.x FE meta image could not be sync

在前两天线上出现一个问题，我三个FE，出现了一个FE挂掉，然后我重启启动不起来，
Doris版本：0.13.11

我备份这个FE元数据后，将这个节点元数据清除掉，使用

--先删除
ALTER SYSTEM DROP FOLLOWER "FE:9010"
--在添加
ALTER SYSTEM ADD FOLLOWER "FE:9010"
删除改节点，然后在使用--helper方式把该节点作为一个新的FE加入到集群中，但是这时候启动会报错，同时会导致Master FE挂掉，具体异常信息如下

repImpl=com.sleepycat.je.rep.impl.RepImpl@68fa4d19 props={REFRESH_VLSN=17921230, PORT=9010, HOSTNAME=172.22.197.238, P_NODETYPE1=ELECTABLE, NODE_NAME=172.22.197.238_9010_1611290318143, P_NODETYPE0=SECONDARY, P_NODENAME1=172.22.197.240_9010_1608972313975, P_PORT1=9010, P_NODENAME0=172.22.197.238_9010_1611290318143, P_PORT0=9010, P_HOSTNAME1=172.22.197.240, GROUP_NAME=PALO_JOURNAL_GROUP, P_HOSTNAME0=172.22.197.238, ENV_DIR=/hdd_data01/doris-meta/bdb, P_NUMPROVIDERS=2}
 at com.sleepycat.je.rep.InsufficientLogException.wrapSelf(InsufficientLogException.java:315) ~[je-7.3.7.jar:7.3.7]
 at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[je-7.3.7.jar:7.3.7]


后来在社区，缪小姐姐及陈明雨大神的协助下，进行了各种尝试定位，认为是在启动的时候元数据同步异常，这个异常可能是因为我当时的load数据任务在同步修改元数据，造成的，后来在凌晨，生产开货完成以后，停掉所有load任务，然后执行删除问题FE节点元数据，然后在重新使用--helper启动，依然报错，最后没办法，尝试将master节点的fe元数据拷贝到问题节点FE，将问题节点FE的元数据目录删除，然后重建，将赋值过来的元数据，拷贝到元数据目录

（临时解决方案）具体步骤：

1.停止所有load任务
2.删除元数据目录，并重建目录
3.从master节点拷贝元数据到问题节点fe（将 fe.conf 中的 metadata_failure_recovery=true 配置项删除，或者设置为 false，这个非常重要）
4.执行 ALTER SYSTEM DROP FOLLOWER  删除改节点
5.在问题节点使用--helper启动服务
6.在mysql下执行 ALTER SYSTEM ADD FOLLOWER 将FE节点从新加入进去
7.启动正常

注意：

1.问题节点：将 fe.conf 中的 metadata_failure_recovery=true 配置项删除，或者设置为 false
2.Master节点启动使用 metadata_failure_recovery=true启动，进行恢复，启动正常以后，将这个配置删除或者设置为false，停掉Master fe，然后在重启，启动完成以后，要确认Master查询，导入是正常的
上述步骤执行完成以后，然后在问题节点在使用--helper方式启动fe，这个时候正常启动，问题解决

重新启动所有Load任务

Today， we meet the same problem.

The slaver exception as following:
```
2021-02-23 19:58:13,992 WARN (UNKNOWN 10.157.102.41_9010_1614081490712(-1)|1) [Catalog.notifyNewFETypeTransfer():2359] notify new FE type transfer: UNKNOWN
2021-02-23 19:58:14,035 WARN (RepNode 10.157.102.41_9010_1614081490712(-1)|45) [Catalog.notifyNewFETypeTransfer():2359] notify new FE type transfer: FOLLOWER
2021-02-23 19:58:14,054 WARN (REPLICA 10.157.102.41_9010_1614081490712(-1)|45) [BDBStateChangeListener.stateChange():59] this node is DETACHED
2021-02-23 19:58:19,386 WARN (UNKNOWN 10.157.102.41_9010_1614081490712(-1)|1) [BDBJEJournal.open():356] catch insufficient log exception. will recover and try again.
com.sleepycat.je.rep.InsufficientLogException: (JE 7.3.7) Environment must be closed, caused by: com.sleepycat.je.rep.InsufficientLogException: Environment invalid because of previous exception: (JE 7.3.7
) 10.157.102.41_9010_1614081490712(-1):/premium_ssd0/484285254228643846/palofe/data/bdb INSUFFICIENT_LOG: Log files at this node are obsolete. Environment is invalid and must be closed. Originally thrown
by HA thread: REPLICA 10.157.102.41_9010_1614081490712(-1) Originally thrown by HA thread: REPLICA 10.157.102.41_9010_1614081490712(-1)refreshVLSN=8,462,234 logProviders=[Node:10.157.102.41_9010_161408149
0712 10.157.102.41:9010 (is member) changeVersion:2 LocalCBVLSN:-1 at:Tue Feb 23 19:58:14 CST 2021 jeVersion:7.3.7
, Node:10.157.102.35_9010_1603177119314 10.157.102.35:9010 (is member) changeVersion:1 LocalCBVLSN:8,462,234 at:Tue Feb 23 19:56:04 CST 2021 jeVersion:7.3.7
] repImpl=com.sleepycat.je.rep.impl.RepImpl@56b517aa props={REFRESH_VLSN=8462234, PORT=9010, HOSTNAME=10.157.102.41, P_NODETYPE1=ELECTABLE, NODE_NAME=10.157.102.41_9010_1614081490712, P_NODETYPE0=ELECTABL
E, P_NODENAME1=10.157.102.35_9010_1603177119314, P_PORT1=9010, P_NODENAME0=10.157.102.41_9010_1614081490712, P_PORT0=9010, P_HOSTNAME1=10.157.102.35, GROUP_NAME=PALO_JOURNAL_GROUP, P_HOSTNAME0=10.157.102.
41, ENV_DIR=/premium_ssd0/484285254228643846/palofe/data/bdb, P_NUMPROVIDERS=2}
        at com.sleepycat.je.rep.InsufficientLogException.wrapSelf(InsufficientLogException.java:315) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkOpen(EnvironmentImpl.java:1775) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Environment.checkOpen(Environment.java:2473) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Environment.getDatabaseNames(Environment.java:2245) ~[je-7.3.7.jar:7.3.7]
        at org.apache.doris.journal.bdbje.BDBEnvironment.getDatabaseNames(BDBEnvironment.java:318) ~[palo-fe.jar:3.4.0]
        at org.apache.doris.journal.bdbje.BDBJEJournal.open(BDBJEJournal.java:329) [palo-fe.jar:3.4.0]
        at org.apache.doris.persist.EditLog.open(EditLog.java:835) [palo-fe.jar:3.4.0]
        at org.apache.doris.catalog.Catalog.initialize(Catalog.java:764) [palo-fe.jar:3.4.0]
        at org.apache.doris.PaloFe.start(PaloFe.java:108) [palo-fe.jar:3.4.0]
        at org.apache.doris.PaloFe.main(PaloFe.java:60) [palo-fe.jar:3.4.0]
Caused by: com.sleepycat.je.rep.InsufficientLogException: Environment invalid because of previous exception: (JE 7.3.7) 10.157.102.41_9010_1614081490712(-1):/premium_ssd0/484285254228643846/palofe/data/bd
b INSUFFICIENT_LOG: Log files at this node are obsolete. Environment is invalid and must be closed. Originally thrown by HA thread: REPLICA 10.157.102.41_9010_1614081490712(-1) Originally thrown by HA thr
ead: REPLICA 10.157.102.41_9010_1614081490712(-1)
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.setupLogRefresh(ReplicaFeederSyncup.java:664) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.getFeederRecord(ReplicaFeederSyncup.java:732) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.findMatchpoint(ReplicaFeederSyncup.java:406) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.execute(ReplicaFeederSyncup.java:151) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.initReplicaLoop(Replica.java:711) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoopInternal(Replica.java:474) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoop(Replica.java:409) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.RepNode.run(RepNode.java:1873) ~[je-7.3.7.jar:7.3.7]
```

--- a/fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java
+++ b/fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java
@@ -1439,13 +1439,13 @@ public class Catalog {
                 response = sb.toString();
             }
 
-			// For http v2, the response body for "/info" api changed from
-			// StorageInfo to StorageInfoV2.
-			// So we need to make it compatible with old api.
+            // For http v2, the response body for "/info" api changed from
+            // StorageInfo to StorageInfoV2.
+            // So we need to make it compatible with old api.
             try {
                 return mapper.readValue(response, StorageInfo.class);
             } catch (Exception e) {
-				// try new response body
+                // try new response body
                 return mapper.readValue(response, StorageInfoV2.class).data;
             }
         } finally {

--- a/fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java
+++ b/fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java
@@ -185,6 +185,7 @@ import org.apache.doris.persist.ReplicaPersistInfo;
 import org.apache.doris.persist.SetReplicaStatusOperationLog;
 import org.apache.doris.persist.Storage;
 import org.apache.doris.persist.StorageInfo;
+import org.apache.doris.persist.StorageInfoV2;
 import org.apache.doris.persist.TableInfo;
 import org.apache.doris.persist.TablePropertyInfo;
 import org.apache.doris.persist.TruncateTableInfo;
@@ -236,12 +237,14 @@ import org.apache.logging.log4j.Logger;
 import org.codehaus.jackson.map.ObjectMapper;
 
 import java.io.BufferedInputStream;
+import java.io.BufferedReader;
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.util.ArrayList;
@@ -1394,7 +1397,7 @@ public class Catalog {
                 MetaHelper.complete(filename, dir);
             }
         } catch (Exception e) {
-            return;
+            throw new IOException(e);
         }
     }
 
@@ -1425,7 +1428,26 @@ public class Catalog {
             connection = (HttpURLConnection) url.openConnection();
             connection.setConnectTimeout(HTTP_TIMEOUT_SECOND * 1000);
             connection.setReadTimeout(HTTP_TIMEOUT_SECOND * 1000);
-            return mapper.readValue(connection.getInputStream(), StorageInfo.class);
+
+            String response;
+            try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
+                String line;
+                StringBuilder sb = new StringBuilder();
+                while ((line = bufferedReader.readLine()) != null) {
+                    sb.append(line);
+                }
+                response = sb.toString();
+            }
+
+			// For http v2, the response body for "/info" api changed from
+			// StorageInfo to StorageInfoV2.
+			// So we need to make it compatible with old api.
+            try {
+                return mapper.readValue(response, StorageInfo.class);
+            } catch (Exception e) {
+				// try new response body
+                return mapper.readValue(response, StorageInfoV2.class).data;
+            }
         } finally {
             if (connection != null) {
                 connection.disconnect();

--- a/fe/fe-core/src/main/java/org/apache/doris/persist/StorageInfoV2.java
+++ b/fe/fe-core/src/main/java/org/apache/doris/persist/StorageInfoV2.java
@@ -1,0 +1,30 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.persist;
+
+/**
+ * This class is wrapper of StorageInfo.
+ * Because for http v2, the response body of "/info" api changed to:
+ * {"msg":"success","code":0,"data":{"clusterID":1464655034,"imageSeq":809779,"editsSeq":0},"count":0}
+ */
+public class StorageInfoV2 {
+    public String msg;
+    public int code;
+    public StorageInfo data;
+    public int count;
+}

